src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,267 @@
+            }
+
+            if (!session.Properties.TryGetProperty(CompletionSource.TriggerSnapshot, out ITextSnapshot triggerSnapshot))
+            {
+                // Need the trigger snapshot to calculate the span when the commit changes to be applied.
+                // It should be inserted into a property bag within GetCompletionContextAsync for each item created by Roslyn.
+                // If not found here, Roslyn should not make a commit.
+                return CommitResultUnhandled;
+            }
+
+            // Commit with completion serivce assumes that null is provided is case of invoke. VS provides '\0' in the case.
+            char? commitChar = typeChar == '\0' ? null : (char?)typeChar;
+            var commitBehavior = Commit(
+                document, completionService, session.TextView, subjectBuffer, 
+                roslynItem, commitChar, triggerSnapshot, serviceRules, filterText, cancellationToken);
+
+            _recentItemsManager.MakeMostRecentItem(roslynItem.DisplayText);
+            return new AsyncCompletionData.CommitResult(isHandled: true, commitBehavior);            
+        }
+
+        private AsyncCompletionData.CommitBehavior Commit(
+            Document document,
+            CompletionService completionService,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            char? commitCharacter,
+            ITextSnapshot triggerSnapshot,
+            CompletionRules rules,
+            string filterText,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer did not provide Edit Access"));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer is editing by someone else."));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
+            {
+                var change = completionService.GetChangeAsync(document, roslynItem, commitCharacter, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                var textChange = change.TextChange;
+
+                var triggerSnapshotSpan = new SnapshotSpan(triggerSnapshot, textChange.Span.ToSpan());
+                var mappedSpan = triggerSnapshotSpan.TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                edit.Replace(mappedSpan.Span, change.TextChange.NewText);
+                var updatedCurrentSnapshot = edit.Apply();
```

Comment:
I know you changed this per @sharwell's feedback, but the use of subjectBuffer.CurrentSnapshot was probably correct: the editor allows edits to happen in response to an edit.Apply(); in that case I'm guessing you want the result after _all_ edits. (And if you're not ready for edits in the middle at all, well, that might be a bigger problem.)

Reply:
OK. Thanks! reverted to subjectBuffer.CurrentSnapshot and added a comment.

---
In reply to: [236134299](https://github.com/dotnet/roslyn/pull/29016#discussion_r236134299) [](ancestors = 236134299)